### PR TITLE
Enable custom script definitions per scan

### PR DESCRIPTION
**Pull request by kconwayatlassian**
Add ability to inject script names and script arguments through the API.

Make showall and vulscan format arguments non-optional because they are
required for correct functioning of the output parsing.

Add NOT VULNERABLE detection to vulscan results by checking the number of
matched systems.

Remove vulscan from default scripts because it is too noisy and not very
accurate. Will need to investigate improving the false-positive rate before
enabling by default. It remains an option for use when requesting custom
scan scripts.

Updated README with more examples.

Adjust all parsers to return empty arrays instead of nil in order to
render empty arrays in the output instead of `null` or `omitempty`.

Add debug logger for nmap calls for development.

### CODE REVIEW

**Review by mikerott — APPROVED** [review]
Can I ask for some additional README docs?  Or perhaps validation of the script and script args in the API yaml (or in code if necessary)?  If it's simply pointing to nmap docs that's good enough.  There was only one example in the README; maybe more?

**Comment by kconwayatlassian** [comment]
> Can I ask for some additional README docs? Or perhaps validation of the script and script args in the API yaml (or in code if necessary)? If it's simply pointing to nmap docs that's good enough. There was only one example in the README; maybe more?

I'll add some in a follow up PR.

File: pkg/scanner/nmap.go [hunk 1]
```diff
@@ -117,39 +125,81 @@ type nmapContainer struct {
...
+	strs := make([]string, 0, 6+len(s.BinArgs))
+	strs = append(
+		strs,
+		s.BinPath,
+		`-sV`,      // Enable version detection of systems.
+		`-oX`, `-`, // Enable XML output mode and send to stdout.
+	)
+	strs = append(strs, s.BinArgs...)
+	if len(s.Scripts) > 0 {
+		strs = append(strs, `--script`, strings.Join(s.Scripts, ","))
```
Comment (ensary) [inline pkg/scanner/nmap.go:192]: Most likely a nit - can we  normalize the path to the script ensure the script(s) really exist on FS level, make sure it is under path where the scripts are supposed to be and not /whetever/path/we/do/not/want ? I understand this comes from internal API, but I still feel not awesome re: running something based on external input.

**Review by ensary — APPROVED** [review]

File: pkg/handlers/v1/scan.go [hunk 2]
```diff
@@ -144,24 +146,37 @@ func fromDomain(source []domain.Finding) []ScanFinding {
...
+		return ScanOutput{}, err
+	}
+	return ScanOutput{Findings: fromDomain(findings)}, nil
+}
+
 // Handle is invoked on each request.
 func (h *Scan) Handle(ctx context.Context, in ScanInput) (interface{}, error) {
-	v, err := h.handle(ctx, in)
+	handler := h.handleScan
+	if len(in.Scripts) > 0 || len(in.ScriptArgs) > 0 {
```
Comment (zlozano) [inline pkg/handlers/v1/scan.go:176]: Question: is it valid to call with script args, but no scripts?

**Review by zlozano — APPROVED** [review]

File: pkg/handlers/v1/scan.go (hunk 2 above)
Comment (kconwayatlassian) [reply pkg/handlers/v1/scan.go:176]: Yes. There are several flags present in "stdlib" modules that might be passed without additional scripts. If no scripts are given then nmap runs a default set of scripts anyway.

**Pull request merged**